### DISCUSSION TITLE
How to check out-ot-gamut pixels?

### DISCUSSION THREAD

**Discussion by hminle** [colour-science/colour#837] [Q&A] [answered]
Hi,

I wonder is there any efficient way to check out-of-gamut pixel values for a specific color space.
For example, I have an image in `ProPhoto RGB` then I convert that image to `Lab` space. After doing some modification, I want to check if any `Lab` values out of `sRGB` gamut?
One way I think of is that I just convert the `Lab` image to `sRGB`:
```python
ILLUMINANT_D50 = colour.CCS_ILLUMINANTS['CIE 1931 2 Degree Standard Observer']['D50']
srgb_colourspace = colour.models.RGB_COLOURSPACE_sRGB.chromatically_adapt(ILLUMINANT_D50)
XYZ = colour.Lab_to_XYZ(Lab_img, 
                            illuminant=ILLUMINANT_D50)
sRGB_img = colour.models.XYZ_to_RGB(XYZ, 
                 srgb_colourspace.whitepoint, 
                 srgb_colourspace.whitepoint,
                 srgb_colourspace.matrix_XYZ_to_RGB)
```
Then I check the `sRGB_img` to see which pixel is out of range [0-1] --> then I call them out-of-sRGB-gamut.

Is my way right?
Is there any other ways more efficient?

Thanks.
P/s: I use `ILLUMINANT_D50` because I think my original image in ProPhoto RGB with D50, then I need to adapt my sRGB to D50 too. Is this necessary or unnecessary?

**Comment by thomasmansencal** [comment, marked as answer]
The easiest way would be to convert from `ProPhoto RGB` to `sRGB` directly using the `colour.RGB_to_RGB` definition and check for negative values. Positive values are not necessarily OOG, as HDR can push them way over 1.

**Reply by hminle** [reply]
Thank @KelSolaar , 
So you mean we use HDR Display? But when we save an image it is still clipped to 1 right? 
And what if we only use Standard Dynamic Range (SDR)? Then OOG colors are the values out side of [0-1] ?

**Reply by thomasmansencal** [reply]
Not necessarily an HDR display but a floating-point format such as .HDR or .EXR that can encode values above 1. If you use a SDR representation then yeah, any RGB value outside [0, 1] is technically OOG.

**Reply by hminle** [reply]
Thank you very much, I got it.

**Reply by hminle** [reply]
Hi @KelSolaar , I have another question. 
A ProPhoto RGB pixel is converted to CIE Lab. Could I check this pixel is outside of sRGB gamut in the Lab space directly?

**Reply by thomasmansencal** [reply]
Not trivially because once in CIE Lab, there are no obvious boundaries left.

**Reply by hminle** [reply]
Thank you 👍

**Reply by hminle** [reply]
> Hi @KelSolaar , I have another question.
> A ProPhoto RGB pixel is converted to CIE Lab. Could I check this pixel is outside of sRGB gamut in the Lab space directly?

Hi @KelSolaar ,

I've just found in the colour documentation this function: 
https://colour.readthedocs.io/en/develop/generated/colour.is_within_mesh_volume.html#colour.is_within_mesh_volume

Could we use this function to check whether a point is out of sRGB gamut in Lab space? 
But I don't know how could we get an entire mesh for sRGB gamut in CIE Lab.

I tried this way:
```python
import numpy as np
import colour

samples = 64
r_values = np.linspace(0, 1, samples)
g_values = np.linspace(0, 1, samples)
b_values = np.linspace(0, 1, samples)

rr, gg, bb = np.meshgrid(r_values, g_values, b_values)
rgb = np.vstack([rr.reshape(-1), gg.reshape(-1) ,bb.reshape(-1)]).T
srgb_colourspace = colour.models.RGB_COLOURSPACE_sRGB

XYZ = colour.models.RGB_to_XYZ(rgb.reshape(-1,3), 
                                 srgb_colourspace.whitepoint, 
                                 srgb_colourspace.whitepoint,
                                 srgb_colourspace.matrix_RGB_to_XYZ)
lab = colour.XYZ_to_Lab(XYZ)
point = np.array([100, -132, 129])
colour.volume.mesh.is_within_mesh_volume(point, lab) # return False
```
Do you think it is correct?

Thanks

**Reply by thomasmansencal** [reply]
Hi @hminle,

I would not recommend using this definition for that: The points in *CIE Lab* space might define a concave volume however the mesh generated by the `colour.volume.mesh.is_within_mesh_volume` definition is based on a convex hull and thus will be larger than the actual *sRGB* colourspace.

You can see that relatively easily here: https://www.colour-science.org:8020/

All the right edges of the *sRGB* colourspace are concave on this image:

![image](https://user-images.githubusercontent.com/99779/132960627-30c01a69-37f9-407c-a45c-d53628ad92c5.png)

**Reply by hminle** [reply]
Thank @KelSolaar I got your points.
